packages/custom-functions-metadata/src/parseTree.ts

```diff
@@ -34,6 +34,7 @@ export interface IFunctionParameter {
   name: string;
   description?: string;
   type: string;
+  subtype?: string;
```

Comment:
Consider something other than "sub" for the name here. It's not really "below". I think a name which is less generic than "type", such as "valueType".

Please provide more background info about the existing types and why there is an issue for backward compatibility. It would help me to understand more and perhaps offer different suggestions here with a better understanding.

Reply:
Hi Adam,
 
Backward compatibility issue: 
When an add-in adopts the new cell value types and use them in the functions.json, it will break the whole custom function registration pipeline on an older client without parser code change, because it's an unknown Enum. This means that no custom function can be used at all. Whereas putting it in a new attribute will make sure that the same metadata can work on both older and newer clients, and a developer can, for example, run different code in the same function after checking the requirement set. Older client will simply ignore the "subtype" attribute.

"Subtype" is used exactly in this sense: it will only be parsed when the "type" is "any", because it's meant to be a refinement of "any" by specifying which specific subtype it is, and it doesn't make sense to add a "subtype" when the "type" is anything other than "any". In that case it will simply be ignored. When "subtype" is invalid, it will also fallback to "any". Please see PR 3058051 for parser code change and relevant tests.
I think if we use "valueType", it wouldn't be clear what the relation between "type" and "valueType" is. It might be confusing to have two seemingly overlapping types there.

Reply:
Really this is where forward-compatibility should have been better considered for unknown types. The parser should have allowed some form of unknown types to be treated as any. Now you're adding a second "type" field to make it for the limitations of the first. 

It seems to me this is specifically targetting support for Cell Value Types, so how about making it just that: "cellValueType". I suppose that when type is something like "number" that cellValueType should be undefined.

What's going to happen when other types are adding in the future. How do older clients handle the fact that there may be new strings they don't recognize.

I don't like "subType" because this type is in now way "below" or "sub" the "type" field.

Reply:
Back/forward compatibility: "any" was introduced alongside rich data in 2019, I believe, and since then there had been no major change to cf schema. We have to work with what we have now.

My major concern for "cellValueType" is that it's not clear from the naming what the relation with "type" is, or which one will take precedence over another when they don't agree. "Subtype" solves these issues since "type" will take precedence, as expected.

Cell Value Types encapsulate all possible content types in a cell, by definition. It's unlikely we will need to extend parameter type again since custom function inputs are values in cells.

Please let me know what you think!

Reply:
Well I've already let you know my perspective. Perhaps @AlexJerabek could weigh in about the naming.

Reply:
I'm confused why we're reworking this for backward compatibility. Can't we say the expanded type support is only supported in Excel version X+ and tie it to a requirement set? That wouldn't be different to how we handle many other aspects of the product. It would also mitigate Adam's concern about the problem in the future.

I'm also confused as to how this actually helps developers. Won't they manually have to change their metadata as you did in the test cases? Isn't that then a breaking change? 

Separate from that, I'm with Adam that "sub" is confusing. As a developer, my instinct would be to provide both type and subtype, but in reality I only need subtype for legacy purposes. Maybe `fallbackType` to make it clear this is only needed if type is invalid?

Tagging @alison-mk and @davidchesnut to weigh in too.

Reply:
Resolving this per discussions via email.